### PR TITLE
chore(bundle-manager): retaining all versions under manifest tags

### PR DESCRIPTION
**Pull request by jordanl17**

### CODE REVIEW

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    ? []
+    : [newVersion]
+
+  return [...withinTtl, ...newVersionEntry]
+}
+
+function getHighestSemverVersion(majorVersions: VersionEntry[]): VersionEntry[] {
+  const highest = majorVersions.toSorted(sortBySemverDescending)
+
+  return highest.length === 1 ? highest : []
```
[the remark this answers is not in the document]
Comment (bjoerge) [reply packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: this seems accurate – are we missing a test case for this?

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+const sortBySemverDescending = (a: VersionEntry, b: VersionEntry): number =>
+  semver.compare(b.version, a.version)
+
+function getMajorVersion(version: string): string {
+  const parsed = semver.parse(version)
+
+  return parsed ? parsed.major.toString() : '0'
+}
+
+function sortByVersion(
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]:
```suggestion
function sortByVersionDesc(
```

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    'version',
+  )
+
+const sortBySemverDescending = (a: VersionEntry, b: VersionEntry): number =>
+  semver.compare(b.version, a.version)
+
+function getMajorVersion(version: string): string {
+  const parsed = semver.parse(version)
+
+  return parsed ? parsed.major.toString() : '0'
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: If parsed returns null it means we're dealing with an invalid semver, which seems like something we should treat like a hard failure?

File: packages/@repo/bundle-manager/src/operations/tagVersion.ts [hunk 4]
```diff
@@ -40,28 +37,13 @@ export function tagVersion(
         ...manifestPackage,
         tags: {
           ...manifestPackage.tags,
-          [tag]: sortAndRemoveOldEntries([tagEntry, ...existingTags]),
+          [tag]: cleanupVersions([tagEntry, ...existingTags], tagEntry),
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/operations/tagVersion.ts]: is it necessary to pass tagEntry to the list of existing versions here? Seems redundant?

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ * Cleans up and sorts version entries with mixed strategy:
+ * - For the new version's major: TTL cleanup (keep multiple versions within grace period)
+ * - For other majors: Keep only highest semver version
+ * @param versions - Array of version entries to clean up
+ * @param newVersion - The version being added (determines which major gets TTL treatment)
+ * @returns Cleaned and sorted array of version entries
+ */
+export function cleanupVersions(
+  versions: VersionEntry[],
+  newVersion?: VersionEntry,
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: We don't seem to ever pass in an undefined newVersion – do we need it?

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+): VersionEntry[] {
+  const uniqueVersions = deduplicateByVersion(versions)
+
+  if (!newVersion) {
+    // Fallback to keeping highest semver per major if no new version specified
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: should we not consider TTL here the same way as for other versions? It seems fine (and expected) that we'll also keep old majors that falls within the buffer window, or maybe I'm missing something?

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+      {timestamp: now - 200, version: '1.1.0'},
+      {timestamp: now - 300, version: '2.0.0'},
+    ]
+
+    const result = cleanupVersions(versions)
+    expect(result).toHaveLength(2) // One per major (highest semver)
+    expect(result.map((v) => v.version)).toEqual(['2.0.0', '1.1.0'])
+  })
+
+  it('applies TTL cleanup for new version major, keeps highest for other majors', () => {
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: Any reason we couldn't keep other majors within the ttl limit here too? Might be missing something, but I don't see the harm, and it would simplify implementation a bit.

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    expect(result.map((v) => v.version)).toEqual([
+      '2.0.0',
+      '2.0.0-beta.1',
+      '2.0.0-alpha.1',
+      '1.0.0',
+    ])
+  })
+})
+
+describe('sortAndCleanupVersions()', () => {
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts:330]: Seems like there's a bit of overlap between `cleanupVersions` tests and `sortAndCleanupVersions`? Would it be better to test the sorting in isolation?

**Review by stipsan — DISMISSED** [review]

**Review by bjoerge — COMMENTED** [review]
Looks good overall – appreciate the extensive tests!

I do wonder what we gain from ignoring ttl when we remove old major versions – could the logic be simplified by applying the same ttl rules to all major versions regardless? Edge case perhaps, but in the event of us tagging 4.50.0 as latest, then tagging 4.50.1 as latest, and then tagging 5.0 as latest and it all happens within the bufffer period, we'd lose 4.50.0 from the list of latest tags when it would be the right one to serve for requests that's not within the safe buffer time for 4.50.1

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts (hunk 3 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: - [x] null parsed should result in failure from malformed semver

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts (hunk 5 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: - [x] cleanupVersions - does it need newVersion

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts (hunk 6 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: - [x] TTL applies across all major versions

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts (hunk 8 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts:330]: I've now broken this down so that we have tests focused only on the sorting, or the filtering, whilst then having a few integration level ones that make sure the combined logic works

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+export function cleanupVersions(
+  existingVersions: VersionEntry[],
+  newVersion: VersionEntry,
+): VersionEntry[] {
+  const allVersions = [newVersion, ...existingVersions]
+  const uniqueVersions = deduplicateByVersion(allVersions)
+
+  uniqueVersions.forEach((entry) => getMajorVersion(entry.version))
+
+  if (!newVersion) {
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: This can never happen now that `newVersion` is required?

File: packages/@repo/bundle-manager/src/helpers/versionUtils.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Cleans up version entries using TTL strategy:
+ * - Keep all versions within TTL window
+ * - Keep highest version per major (if not already kept via TTL)
+ */
+export function cleanupVersions(
+  existingVersions: VersionEntry[],
+  newVersion: VersionEntry,
+): VersionEntry[] {
+  const allVersions = [newVersion, ...existingVersions]
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/versionUtils.ts]: Given that the only thing we do with `newVersion` here is to prepend it, could that be the responsibility of the caller? Seems a lot easier to reason about this function if it just operates on the given list of versions.

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+      major1HigherVersionInTTL,
+      major1LowerVersionOutsideTTL,
+      major2OnlyVersionOutsideTTL,
+    ]
+
+    const newVersion = {timestamp: now, version: '3.0.0'}
+    const result = cleanupVersions(existingVersions, newVersion)
+
+    expect(result).toHaveLength(3)
+    expect(new Set(result.map((v) => v.version))).toEqual(new Set(['3.0.0', '1.3.0', '2.1.0']))
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: I find it a bit hard to understand why we're using `Set` here. Is it to make the order not matter? If so, that doesn't seem necessary as the returned list has a defined sort order. The problem with using `Set` is that it remove duplicates, which is not something we intentionally test for here.

```suggestion
    expect(result.map((v) => v.version)).toEqual(['3.0.0', '2.1.0', '1.3.0'])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+  })
+
+  it('keeps highest semver per major even when outside TTL if major also has TTL versions', () => {
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: Found it a bit hard to unpack the rationale for the behavior here, and think future readers could benefit from a comment outlining why we're testing this case :)

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+      ...major1VersionsInTTL,
+      ...major2VersionsOutsideTTL,
+      ...major2VersionsInTTL,
+    ]
+
+    const newVersion = {timestamp: now, version: '2.3.0'}
+    const result = cleanupVersions(existingVersions, newVersion)
+
+    expect(result).toHaveLength(4)
+    expect(new Set(result.map((v) => v.version))).toEqual(
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]:
```suggestion
    expect(result.map((v) => v.version)).toEqual(['2.3.0', '2.2.0', '2.1.0', '1.2.0'])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+    ]
+
+    const newVersionOutsideTTL = {
+      timestamp: now - STALE_TAGS_EXPIRY_SECONDS - 100,
+      version: '2.3.0',
+    }
+    const result = cleanupVersions(versionsInTTL, newVersionOutsideTTL)
+
+    expect(result).toHaveLength(3)
+    expect(new Set(result.map((v) => v.version))).toEqual(new Set(['2.3.0', '2.1.0', '2.0.0']))
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]:
```suggestion
    expect(result.map((v) => v.version)).toEqual(['2.3.0', '2.1.0', '2.0.0'])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+    ]
+
+    const existingVersions = [...major1VersionsInTTL, ...major2VersionsOutsideTTL]
+    const newVersionInDifferentMajor = {timestamp: now, version: '3.0.0'}
+    const result = cleanupVersions(existingVersions, newVersionInDifferentMajor)
+
+    expect(result).toHaveLength(4)
+    expect(new Set(result.map((v) => v.version))).toEqual(
+      new Set(['3.0.0', '2.1.0', '1.1.0', '1.0.0']),
+    )
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]:
```suggestion
	expect(result.map((v) => v.version)).toEqual(['3.0.0', '2.1.0', '1.1.0', '1.0.0'])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+    const regularVersionInTTL = {timestamp: now - 300, version: '1.0.0'}
+
+    const existingVersions = [...preReleaseVersionsInTTL, regularVersionInTTL]
+    const newStableVersion = {timestamp: now, version: '2.0.0'}
+    const result = cleanupVersions(existingVersions, newStableVersion)
+
+    expect(result).toHaveLength(4)
+    expect(new Set(result.map((v) => v.version))).toEqual(
+      new Set(['2.0.0', '2.0.0-beta.1', '2.0.0-alpha.1', '1.0.0']),
+    )
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]:
```suggestion
	expect(result.map((v) => v.version)).toEqual([
      '2.0.0',
      '2.0.0-beta.1',
      '2.0.0-alpha.1',
      '1.0.0',
    ])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    ]
+
+    const newVersion = {timestamp: now, version: '3.0.0'}
+    const result = cleanupVersions(existingVersions, newVersion)
+
+    expect(result).toHaveLength(4)
+    expect(new Set(result.map((v) => v.version))).toEqual(
+      new Set(['3.0.0', '1.3.0', '1.2.0', '2.1.0']),
+    )
+  })
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: In this case it seems pertinent that we retain two entries of the same version?

```suggestion
    expect(result.map((v) => v.version)).toEqual(['3.0.0', '2.1.0', '1.3.0', '1.2.0'])
```

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts (hunk 12 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: Do you think this reads better? "should keep all TTL versions plus highest outside-TTL version per major" - I'm trying to reword the test cases to make them clearer

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts (hunk 17 above)
Comment (jordanl17) [reply packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: Why would we retain 2 of the same semver? We'd surely only ever want the most recent because when returning these when the bundle is requested, we'll sort by most recent

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts (hunk 12 above)
Comment (bjoerge) [reply packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: Much better! Thank you

File: packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts (hunk 17 above)
Comment (bjoerge) [reply packages/@repo/bundle-manager/src/helpers/__test__/versionUtils.test.ts]: seems like I might have misunderstood what this test was about. What you say here makes sense!

File: packages/@repo/bundle-manager/src/operations/__test__/tagVersion.test.ts [hunk 18]
```diff
@@ -105,6 +105,35 @@ describe('tagVersion()', () => {
...
+
+    const tag1 = {timestamp: currentUnixTime() - 20, version: '1.2.4' as const}
+    const tag2 = {timestamp: currentUnixTime() - 30, version: '2.1.0' as const}
+
+    const result = tagVersion(tagVersion(manifest, 'stable', tag1), 'stable', tag2)
+
+    expect(result).toEqual({
+      tags: {
+        stable: [
+          {timestamp: currentUnixTime() - 30, version: '2.1.0'},
```
Comment (bjoerge) [inline packages/@repo/bundle-manager/src/operations/__test__/tagVersion.test.ts]: this would fail if we hop to the next second in between tag1.timestamp is set and this line is invoked – better to refer to either `tag2.timestamp` or even change the whole array to `[tag2, tag1]` as done in the other tests

Comment (jordanl17) [reply packages/@repo/bundle-manager/src/operations/__test__/tagVersion.test.ts]: Have done some test assertion housekeeping [here](https://github.com/sanity-io/sanity/pull/10063/commits/54eda2f59ca94d91601b0c84f83c9816a65685ea)
